src/compile/model.ts

```diff
@@ -268,7 +270,14 @@ export abstract class Model {
   public abstract assembleSelectionSignals(): any[];
 
   public abstract assembleSelectionData(data: VgData[]): VgData[];
-  public abstract assembleData(): VgData[];
+  public assembleData(): VgData[] {
+     if (!this.parent) {
+      // only assemble data in the root
+      optimizeDataflow(this.component.data);
```

Comment:
I moved it to compile.ts. I don't want to put it into parse because then I hate to check whether we are in the root again.